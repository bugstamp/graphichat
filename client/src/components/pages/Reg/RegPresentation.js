import React, { memo } from 'react';
import { map } from 'lodash';

import Typography from '@material-ui/core/Typography';

import BrandTitle from '../Login/BrandTitle';

import {
  RegPresentationWrapper,
  SubTitle,
  SubTitleWord,
  Enjoy,
} from './styled';
import { isEven } from '../../../helpers';

export const words = ['Simple.', 'Fun.', 'Fast.', 'Useful.', 'Powerful.'];

export const getWordPropsByIndex = (index) => {
  const num = index + 1;
  const order = isEven(num) ? 'even' : 'odd';
  const delay = (num / words.length) + 1;

  return { order, delay };
};

const RegPresentation = () => (
  <RegPresentationWrapper>
    <BrandTitle />
    <SubTitle>
      {
        map(words, (word, index) => {
          const wordProps = getWordPropsByIndex(index);

          return (
            <SubTitleWord
              key={word}
              component="p"
              variant="h4"
              align="center"
              {...wordProps}
            >
              {word}
            </SubTitleWord>
          );
        })
      }
    </SubTitle>
    <Enjoy>
      <Typography
        component="p"
        variant="h4"
        align="center"
        color="inherit"
      >
        Enjoy!
      </Typography>
    </Enjoy>
  </RegPresentationWrapper>
);

export { RegPresentation };

export default memo(RegPresentation);
