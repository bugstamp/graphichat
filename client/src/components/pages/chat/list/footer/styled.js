import styled from 'styled-components';

import { getSpacing } from '../../../../../styles';

export const ChatListFooterStyled = styled.div`
  display: flex;
  justify-content: center;
  padding: ${getSpacing(2)} ${getSpacing(1)};

  && {
    button {
      width: 50%;
    }
  }
`;

export default null;
