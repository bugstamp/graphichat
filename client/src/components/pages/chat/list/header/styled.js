import styled from 'styled-components';
import { position } from 'polished';

import IconButton from '@material-ui/core/IconButton';

import { getSpacing } from '../../../../../styles';

export const ChatListHeaderStyled = styled.div`
  display: flex;
  flex-flow: column;
  padding: ${getSpacing(2)} ${getSpacing(1)};
`;

export const HeaderTitle = styled.div`
  position: relative;
  margin-bottom: ${getSpacing(1)};
`;

export const AddButton = styled(IconButton)`
  && {
    ${position('absolute', 0, 0, null, null)};
    padding: ${getSpacing(0.5)};
  }
`;
