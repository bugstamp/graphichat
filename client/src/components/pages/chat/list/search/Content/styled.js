import styled from 'styled-components';

import DialogContent from '@material-ui/core/DialogContent';

import { SearchBoxStyled } from '../../../../../common/SearchBox/styled';
import { getStyledProps, getSpacing } from '../../../../../../styles';

export const SearchDialogContentStyled = styled(DialogContent)`
  && {
    flex: 1 auto;
    display: flex;
    flex-flow: column;
    padding: ${getSpacing(2)} ${getSpacing(3)};
    padding-bottom: 0;

    ${SearchBoxStyled} {
      margin-bottom: ${getSpacing(2)}
    }

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const spacing = getStyledProps('theme.spacing')(props);
    const smDown = breakpoints.down('sm');

    return `
      ${SearchBoxStyled} {
        margin-bottom: ${getSpacing(1)}
      }
      ${smDown} {
        padding: ${spacing(1)}px ${spacing(2)}px;
      }
    `;
  }}
  }
`;

export default null;
