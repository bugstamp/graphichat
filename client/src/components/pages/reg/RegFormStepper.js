import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { map } from 'lodash';

import Stepper from '@material-ui/core/Stepper';
import Step from '@material-ui/core/Step';

import StepLabelIcon from './StepLabelIcon';
import { StepLabelStyled as StepLabel } from './styled';

export const steps = [
  'Create your account',
  'Tell us about yourself',
];

const RegFormStepper = ({ activeStep }) => (
  <Stepper alternativeLabel activeStep={activeStep}>
    {
      map(steps, (label, index) => {
        const step = index + 1;
        const isActive = activeStep === index;
        const isCompleted = index < activeStep;
        const iconColor = (!isCompleted && !isActive)
          ? 'disabled'
          : 'primary';

        return (
          <Step
            key={index}
            active={isActive}
            completed={isCompleted}
          >
            <StepLabel
              icon={(
                <StepLabelIcon
                  isCompleted={isCompleted}
                  step={step}
                  iconColor={iconColor}
                />
              )}
            >
              <span className="step">{`Step ${step}.`}</span>
              <p>{label}</p>
            </StepLabel>
          </Step>
        );
      })
    }
  </Stepper>
);

RegFormStepper.propTypes = {
  activeStep: PropTypes.number.isRequired,
};

export { RegFormStepper };

export default memo(RegFormStepper);
