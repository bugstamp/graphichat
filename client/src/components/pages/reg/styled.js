import styled, { keyframes } from 'styled-components';
import { backgrounds } from 'polished';
import {
  fadeIn,
  fadeInRight,
  fadeInLeft,
  hinge,
} from 'react-animations';

import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import StepLabel from '@material-ui/core/StepLabel';

import { getStyledProps, getSpacing } from '../../../styles';
import bgImage from '../../../assets/images/reg-bg__1920_65.jpg';

const fadeInAnimation = keyframes`${fadeIn}`;
const fadeInRightAnimation = keyframes`${fadeInRight}`;
const fadeInLeftAnimation = keyframes`${fadeInLeft}`;
const hingeAnimation = keyframes`${hinge}`;

export const RegWrapper = styled(Grid)`
  && {
    flex: 1 auto;
    display: flex;
    flex-flow: row nowrap;
    position: relative;
    ${backgrounds(`url(${bgImage})`, 'no-repeat')}
    background-size: cover;
    background-position: center;
    overflow: hidden;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const smDown = breakpoints.down('sm');

    return `
      ${smDown} {
        flex-flow: column;
      }
    `;
  }}
  }
`;

export const RegFormWrapper = styled.div`
  flex: 1 40%;
  max-height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  opacity: 0;
  animation: 1.5s ${fadeInAnimation};
  animation-delay: 1s;
  animation-fill-mode: forwards;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const smDown = breakpoints.down('sm');

    return `
      ${smDown} {
        animation-play-state: paused;
        opacity: 1;
      }
    `;
  }}
`;

export const RegPresentationWrapper = styled.div`
  flex: 1 60%;
  display: flex;
  flex-flow: column;
  align-items: center;
  justify-content: center;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const smDown = breakpoints.down('sm');

    return `
      ${smDown} {
        display: none;
      }
    `;
  }}
`;

export const SubTitle = styled.div`
  color: #fff;
  text-align: center;
`;

export const SubTitleWord = styled(Typography)`
  && {
    display: inline;
    margin-right: ${getSpacing(1)};
    font-weight: bold;
    opacity: 0;
    animation: 1s ${({ order }) => (order === 'even' ? fadeInLeftAnimation : fadeInRightAnimation)};
    animation-delay: ${({ delay }) => `${delay}s`};
    animation-fill-mode: forwards;
  }
`;

export const Enjoy = styled.div`
  margin-top: ${getSpacing(3)};
  color: #fff;
  opacity: 0;
  animation: 2s ${hingeAnimation};
  animation-delay: 2s;
  animation-fill-mode: forwards;

  p {
    font-weight: bold;
  }
`;

export const FormWrapper = styled(Paper)`
  && {
    width: 100%;
    max-height: 100%;
    max-width: 375px;
    min-width: 320px;
    position: relative;
    margin: auto 0;
    padding: ${getSpacing(4)} ${getSpacing(3)};
    overflow-y: auto;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const spacing = getStyledProps('theme.spacing')(props);
    const smDown = breakpoints.down('sm');
    const xsDown = breakpoints.down('xs');

    return `
      ${smDown} {
        max-width: 100%;
        height: 100%;
        border-radius: 0;
        padding: ${spacing(4)}px ${spacing(20)}px;
      }
      ${xsDown} {
        padding: ${spacing(2)}px;
      }
    `;
  }}
  }
`;

export const FormHeader = styled(Typography)`
  && {
    width: 100%;
    position: relative;
  }
`;

export const FormInfo = styled(Typography)`
  && {
    text-align: center;
  }
`;

export const FormFooter = styled(Typography)`
  && {
    width: 100%;
    margin-top: ${getSpacing(2)};
  }
`;

export const StepLabelStyled = styled(StepLabel)`
  p {
    width: 100%;
  }
`;
