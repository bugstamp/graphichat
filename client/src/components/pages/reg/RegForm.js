import React, { useCallback, useMemo } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { useMutation } from '@apollo/react-hooks';

import Form from '../../common/Form/Form';
import SocialMedia from '../../common/SocialMedia/SocialMedia';
import RegFormStepper from './RegFormStepper';
import TopProgressLine from '../../common/TopProgressLine';

import {
  FormWrapper,
  FormHeader,
  FormInfo,
  FormFooter,
} from './styled';

import gql from '../../../gql';

const {
  SIGN_UP,
  SIGN_UP_COMPLETION,
  SIGN_UP_BY_SOCIAL,
} = gql;

const RegForm = (props) => {
  const {
    userId,
    activeStep,
    isCompleted,
    onSuccess,
    onError,
  } = props;
  const asyncFields = useMemo(() => ['username', 'email'], []);
  const [signUp, signUpResult] = useMutation(SIGN_UP, {
    onCompleted: onSuccess('signUp'),
  });
  const [signUpCompletion, signUpCompletionResult] = useMutation(SIGN_UP_COMPLETION, {
    onCompleted: onSuccess('signUpCompletion'),
  });
  const [signUpBySocial, signUpBySocialResult] = useMutation(SIGN_UP_BY_SOCIAL, {
    onCompleted: onSuccess('signUpBySocial'),
    onError,
  });
  const loading = signUpResult.loading || signUpCompletionResult.loading;

  const handleSignUpCompletion = useCallback(({ variables }) => {
    signUpCompletion({ variables: { ...variables, id: userId } });
  }, [userId, signUpCompletion]);

  return (
    <FormWrapper elevation={8}>
      <TopProgressLine loading={loading} />
      <FormHeader variant="h1" color="primary" align="center" gutterBottom>
        <RegFormStepper activeStep={activeStep} />
      </FormHeader>
      <Choose>
        <When condition={isCompleted}>
          <FormInfo>
            Your account had been successfuly completed.Check your email and confirm registration.
          </FormInfo>
        </When>
        <When condition={activeStep === 0}>
          <Form
            formId="signUpStepOne"
            asyncFields={asyncFields}
            mutation={signUp}
            result={signUpResult}
            submitButtonText="Confirm"
          />
          <SocialMedia
            mutation={signUpBySocial}
            result={signUpBySocialResult}
          />
        </When>
        <When condition={activeStep === 1}>
          <Form
            formId="signUpStepTwo"
            mutation={handleSignUpCompletion}
            result={signUpCompletionResult}
            submitButtonText="Confirm"
          />
        </When>
        <Otherwise>
          {null}
        </Otherwise>
      </Choose>
      <FormFooter align="center">
        <Link to="/login">I already have an account</Link>
      </FormFooter>
    </FormWrapper>
  );
};

RegForm.defaultProps = {
  userId: null,
};
RegForm.propTypes = {
  userId: PropTypes.string,
  activeStep: PropTypes.number.isRequired,
  isCompleted: PropTypes.bool.isRequired,
  onSuccess: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired,
};

export default RegForm;
