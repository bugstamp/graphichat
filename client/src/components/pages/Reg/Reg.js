import React, {
  useState,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';
import queryString from 'query-string';

import RegContainer from '../../containers/RegContainer';
import RegForm from './RegForm';
import RegPresentation from './RegPresentation';

import withNotification from '../../common/HOC/withNotification';
import storage, { checkToken } from '../../../storage';

import { RegWrapper, RegFormWrapper } from './styled';

const Reg = (props) => {
  const {
    location: { search },
    history,
    toggleNotification,
  } = props;
  const [activeStep, setActiveStep] = useState(0);
  const [completed, setStatus] = useState(false);
  const steps = [
    'Create your account',
    'Tell us about yourself',
  ];

  const historyPush = path => history.push(path);

  function handleSetActiveStep(step) {
    setActiveStep(step);
  }

  function handleSetStatus(status) {
    setStatus(status);
  }

  function handleSuccess({ token, refreshToken }) {
    if (token && refreshToken) {
      storage.setTokens(token, refreshToken);
      historyPush('/');
    } else {
      handleSetStatus(true);
    }
  }

  function handleError(e) {
    if (e.graphQLErrors) {
      const { graphQLErrors } = e;
      const { message } = graphQLErrors[0];

      toggleNotification(message);
    }
  }

  useEffect(() => {
    const { token } = queryString.parse(search);

    if (token) {
      try {
        const { regStatus } = checkToken(token, true);

        if (regStatus) {
          handleSetActiveStep(1);
        } else {
          historyPush('/reg');
        }
      } catch (e) {
        throw e;
      }
    }
  });

  return (
    <RegContainer
      signUpProps={{
        onCompleted: ({ signUp }) => handleSuccess(signUp),
      }}
      signUpCompletionProps={{
        onCompleted: ({ signUpCompletion }) => handleSuccess(signUpCompletion),
      }}
      signUpBySocialProps={{
        onCompleted: ({ signUpBySocial }) => handleSuccess(signUpBySocial),
        onError: handleError,
      }}
    >
      {({
        signUpAsyncValidationUsername,
        signUpAsyncValidationEmail,
        signUp,
        signUpCompletion,
        signUpBySocial,
      }) => (
        <RegWrapper container>
          <RegPresentation />
          <RegFormWrapper>
            <RegForm
              steps={steps}
              activeStep={activeStep}
              completed={completed}
              signUpAsyncValidationUsername={signUpAsyncValidationUsername}
              signUpAsyncValidationEmail={signUpAsyncValidationEmail}
              signUp={signUp}
              signUpCompletion={signUpCompletion}
              signUpBySocial={signUpBySocial}
            />
          </RegFormWrapper>
        </RegWrapper>
      )}
    </RegContainer>
  );
};

Reg.propTypes = {
  location: PropTypes.shape({
    search: PropTypes.string,
  }).isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired,
  toggleNotification: PropTypes.func.isRequired,
};

export default withNotification(Reg);
