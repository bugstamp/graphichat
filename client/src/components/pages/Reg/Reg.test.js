import config from 'config';
import React from 'react';
import { act } from 'react-dom/test-utils';
import wait from 'waait';
import jwt from 'jsonwebtoken';

import { Reg } from './Reg';
import RegPresentation from './RegPresentation';
import { RegWrapper } from './styled';

import { mountMockedProvider } from '../../../__mocks__/mockedProvider';
import { checkSessionExpirationMock } from '../../../__mocks__/mockedQueries';
import { tokens } from '../../../__mocks__/mockedQueryData';
import historyMock, { historyPushMock } from '../../../__mocks__/historyMock';
import { fixReactFacebookLoginBug } from '../../../__mocks__/testUtils';

import storage from '../../../storage';

const { tokenSecrets } = config;
const defaultMocks = [checkSessionExpirationMock];

describe('Reg', () => {
  fixReactFacebookLoginBug();

  let locationMock = {
    search: '',
  };
  const toggleNotificationMock = jest.fn();
  const mockToken = {
    data: {
      regStatus: 'COMPLETED',
    },
  };
  const token = jwt.sign(mockToken, tokenSecrets.token);

  const mountWrapper = (mocks = defaultMocks, options = {}) => mountMockedProvider(
    <Reg
      location={locationMock}
      history={historyMock}
      toggleNotification={toggleNotificationMock}
    />,
    mocks,
    options,
  );

  afterEach(() => {
    jest.clearAllMocks();
    storage.removeTokens();
  });

  test('should mount without errors', () => {
    const wrapper = mountWrapper();

    expect(wrapper.find('Reg')).toHaveLength(1);
  });
  test('RegWrapper should have container prop', () => {
    const wrapper = mountWrapper();
    const RegWrapperTest = wrapper.find(RegWrapper);

    expect(RegWrapperTest).toHaveLength(1);
    expect(RegWrapperTest.prop('container')).toBeTruthy();
  });
  test('should have RegPresentation', () => {
    const wrapper = mountWrapper();

    expect(wrapper.find(RegPresentation)).toHaveLength(1);
  });
  test('RegForm should have correct props', () => {
    const wrapper = mountWrapper();

    expect(wrapper.find('RegForm')).toHaveLength(1);

    const props = wrapper.find('RegForm').props();
    expect(props).toMatchObject({
      activeStep: 0,
      isCompleted: false,
    });
    expect(props.onSuccess).toEqual(expect.any(Function));
    expect(props.onError).toEqual(expect.any(Function));
  });

  describe('componentDidMount', () => {
    test('should passed first step when mounted with token in the location search string', async () => {
      await act(async () => {
        locationMock = {
          search: `?token=${token}`,
        };
        const wrapper = mountWrapper();

        await wait();
        wrapper.update();

        expect(wrapper.find('RegForm').prop('activeStep')).toEqual(1);
      });
    });
    test('should redirect to reg page when mounted with invalid token in the location search string', async () => {
      await act(async () => {
        locationMock = {
          search: '?token=234324',
        };
        mountWrapper();
        await wait();

        expect(historyPushMock).toBeCalledWith('/reg');
      });
    });
  });
  describe('onSuccess callback', () => {
    test('should save tokens in the storage and redirect to main page when called with tokens', async () => {
      await act(async () => {
        const wrapper = mountWrapper();
        const onSuccess = wrapper.find('RegForm').prop('onSuccess');
        onSuccess('signUp')({ signUp: tokens });

        await wait();

        expect(storage.getTokens()).toEqual(tokens);
        expect(historyPushMock).toBeCalledWith('/');
      });
    });
    test('should set completed reg status when called without tokens', async () => {
      await act(async () => {
        const wrapper = mountWrapper();
        const onSuccess = wrapper.find('RegForm').prop('onSuccess');
        onSuccess('signUp')({ signUp: true });

        await wait();
        wrapper.update();

        expect(wrapper.find('RegForm').prop('isCompleted')).toEqual(true);
      });
    });
  });
  test('onError callback should invoke toggleNotification with error message', async () => {
    const wrapper = mountWrapper();
    const message = 'Invalid field';
    const onError = wrapper.find('RegForm').prop('onError');

    expect(onError).toEqual(expect.any(Function));

    await act(async () => {
      onError({ graphQLErrors: [{ message, data: { invalidField: null } }] });

      expect(toggleNotificationMock).toBeCalledWith(message);
    });
  });
});
