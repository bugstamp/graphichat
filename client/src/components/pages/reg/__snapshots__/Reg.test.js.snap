// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reg should match to snapshot 1`] = `
<styled__RegWrapper
  container={true}
>
  <RegPresentation />
  <styled__RegFormWrapper>
    <RegForm
      activeStep={0}
      isCompleted={false}
      onError={[Function]}
      onSuccess={[Function]}
      userId={null}
    />
  </styled__RegFormWrapper>
</styled__RegWrapper>
`;
